ntoskrnl/io/pnpmgr/pnpirp.c

```diff
@@ -306,3 +306,61 @@ PiIrpQueryPnPDeviceState(
 
     return status;
 }
+
+// IRP_MN_QUERY_CAPABILITIES (0x09)
+NTSTATUS
+PiIrpQueryPnPDeviceCapabilities(
+    _In_ PDEVICE_NODE DeviceNode,
+    _Out_ PDEVICE_CAPABILITIES DeviceCaps)
+{
+    PAGED_CODE();
+
+    ASSERT(DeviceNode);
+
+    RtlZeroMemory(DeviceCaps, sizeof(DEVICE_CAPABILITIES));
+    DeviceCaps->Size = sizeof(DEVICE_CAPABILITIES);
+    DeviceCaps->Version = 1;
+    DeviceCaps->Address = -1;
+    DeviceCaps->UINumber = -1;
+
+    IO_STACK_LOCATION stack = {
+        .MajorFunction = IRP_MJ_PNP,
+        .MinorFunction = IRP_MN_QUERY_CAPABILITIES
+    };
+    stack.Parameters.DeviceCapabilities.Capabilities = DeviceCaps;
+
+    PVOID Dummy;
+    return IopSynchronousCall(DeviceNode->PhysicalDeviceObject, &stack, &Dummy);
+}
+
+// IRP_MN_QUERY_ID (0x13)
+NTSTATUS
+PiIrpQueryPnPDeviceId(
+    _In_ PDEVICE_NODE DeviceNode,
+    _In_ BUS_QUERY_ID_TYPE IdType,
+    _Out_ PWCHAR *Id)
```

Comment:
Is it a string that is guaranteed to be always NULL-terminated?

Reply:
Well, this is what MSDN says:
If a driver returns ID(s) in response to this IRP, it allocates a WCHAR structure from paged pool to contain the ID(s). The PnP manager frees the structure when it is no longer needed.

Reply:
Of course. But if it doesn't also return the size of that string allocation, this certainly means that this string has to be null-terminated. Otherwise nobody would know where to stop its parsing.

Reply:
What do you suggest?

Reply:
If it's just one null-terminated string, then `PWSTR* Id`; if it's a multi-string, then `PZZWSTR* Id` (if there can be many, then `Ids` in plural maybe?)

Reply:
`IRP_MN_QUERY_ID` returns different strings depending on the ID type requested. So let's just keep it `PWSTR`.